src/api-binder.ts

```diff
@@ -443,309 +528,259 @@ export class APIBinder {
-					} else {
-						throw err;
-					}
-				}
-				opts.registered_at = Date.now();
-			} else if (opts.provisioningApiKey != null) {
-				log.debug(
-					'Device is registered but we still have an apiKey, attempting key exchange',
-				);
-				device = await this.exchangeKeyAndGetDevice(opts);
-			}
-
-			if (!device) {
-				// TODO: Type this?
-				throw new Error(`Failed to provision device!`);
-			}
-			const { id } = device;
-			if (!this.balenaApi) {
-				throw new InternalInconsistencyError(
-					'Attempting to provision a device without an initialized API client',
-				);
-			}
-			this.balenaApi.passthrough.headers.Authorization = `Bearer ${opts.deviceApiKey}`;
-
-			const configToUpdate = {
-				registered_at: opts.registered_at,
-				deviceId: id,
-				apiKey: null,
-			};
-			await config.set(configToUpdate);
-			eventTracker.track('Device bootstrap success');
-		}
-
-		// Now check if we need to pin the device
-		const pinValue = await config.get('pinDevice');
+async function provisionDevice() {
+	if (balenaApi == null) {
+		throw new Error(
+			'Trying to provision a device without initializing API client',
+		);
+	}
 
-		if (pinValue != null) {
-			if (pinValue.app == null || pinValue.commit == null) {
-				log.error(
-					`Malformed pinDevice fields in supervisor database: ${pinValue}`,
-				);
-				return;
-			}
-			log.info('Attempting to pin device to preloaded release...');
-			return this.pinDevice(pinValue);
-		}
+	const conf = await config.getMany([
+		'apiKey',
+		'bootstrapRetryDelay',
+		'pinDevice',
+		'provisioned',
+	]);
+
+	if (!conf.provisioned || conf.apiKey != null || conf.pinDevice != null) {
+		await provisionOrRetry(conf.bootstrapRetryDelay);
+		globalEventBus.getInstance().emit('deviceProvisioned');
+		return;
 	}
 
-	private async provisionOrRetry(retryDelay: number): Promise<void> {
-		eventTracker.track('Device bootstrap');
-		try {
-			await this.provision();
-		} catch (e) {
-			eventTracker.track(`Device bootstrap failed, retrying`, {
-				error: e,
-				delay: retryDelay,
-			});
-			await Bluebird.delay(retryDelay);
-			return this.provisionOrRetry(retryDelay);
-		}
+	return conf;
+}
+
+async function reportInitialName(
+	deviceId: number,
+	name: string,
+): Promise<void> {
+	if (balenaApi == null) {
+		throw new InternalInconsistencyError(
+			`Attempt to set an initial device name without an API client`,
+		);
 	}
 
-	private async provisionDevice() {
-		if (this.balenaApi == null) {
-			throw new Error(
-				'Trying to provision a device without initializing API client',
-			);
-		}
-		const conf = await config.getMany([
-			'provisioned',
-			'bootstrapRetryDelay',
-			'apiKey',
-			'pinDevice',
-		]);
-
-		if (!conf.provisioned || conf.apiKey != null || conf.pinDevice != null) {
-			await this.provisionOrRetry(conf.bootstrapRetryDelay);
-			globalEventBus.getInstance().emit('deviceProvisioned');
-			return;
-		}
+	await balenaApi.patch({
```

Comment:
Let's wrap this in a  try/catch because we don't want to break if for some reason we couldn't set the device name